### DISCUSSION TITLE
[Item] A slightly modified version of annotation collection by color

### DISCUSSION THREAD

**Discussion by LynnXie00** [windingwind/zotero-better-notes#406] [Note Templates]
**2024-10-10

Guys, I updated the scripts. With the Zotero 7 and Better Notes, for some reason, the script can't import annotations that contain images anymore. (I'm not a programmer and that's all I can do lol.... So you probably need to import images into Zotero notes manually. But the other part works fine for me. **  

This code is modified based on later versions in this thread
https://github.com/windingwind/zotero-better-notes/discussions/406#discussioncomment-6887168

**Type:** Item
**Description:** This template maps certain colors with certain meanings in annotations.
I use gray for purpose, yellow for general, red for attention, green for hypotheses, blue for theory, purple for method, magenta for results, and orange for contribution. 
You may change this code to change your own color style.

```
		const colors = {
			"#aaaaaa": "💡Purpose", // Grey
			"#2ea8e5": " 📚Theory", // Blue
			"#5fb236": "❗Hypotheses", // Green
			"#a28ae5": " ⛏️Method", // Purple
			"#e56eee": "📝Results", // Magenta
			"#f19837": "👍Contribution", // Orange
			"#ff6666": "⁉️ Attention", // Red
		        "#ffd400": "General ",  // Yellow, if you don't collect general highlight, just delete it

		}
```

**How to use it**: 
![image](https://github.com/user-attachments/assets/bd4ad777-6162-4e57-a2c6-086f039011c1)
![image](https://github.com/user-attachments/assets/142c9283-b8d0-4fdd-a128-1c279b91c489)

**Showcase result**: 
![image](https://github.com/user-attachments/assets/859d8e0c-3019-446d-a82e-8b2a1c64b3f3)

**Template Contents:**  

```${
	{
		async function getAnnotationsByColor(_item, color) {
			const annots = _item
				.getAnnotations()
				.filter((item) => item.annotationColor === color);
			if (annots.length === 0) {
				return "";
			}
			return Zotero.BetterNotes.api.convert.annotations2html(annots, {
				noteItem: targetNoteItem,
			});
		}

		const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
			i.isPDFAttachment()
		);

		const creators = topItem.getCreators();
		let content = "";
		const year = topItem.getField("year");
		if (creators.length === 0) {
			content = "no author";
		} else if (creators.length === 1) {
			content = `${creators[0].lastName}, ${year}`;
		} else {
			content = `${creators[0].lastName} etal., ${year}`;
		}

		let res = `<h1>${content}</h1> <h2>${topItem.getField("title")}</h2>`;

		const colors = {
			"#aaaaaa": "💡Purpose", // Grey
			"#2ea8e5": " 📚Theory", // Blue
			"#5fb236": "❗Hypotheses", // Green
			"#a28ae5": " ⛏️Method", // Purple
			"#e56eee": "📝Results", // Magenta
			"#f19837": "👍Contribution", // Orange
			"#ff6666": "⁉️ Attention", // Red
                        "#ffd400": "General ",  // Yellow, if you don't collect general highlight, just delete it
		};
		for (const attachment of attachments) {
			res += `<h2>📄For Document:${attachment.attachmentFilename}</h2>`;
			for (const color in colors) {
				if (Object.hasOwnProperty.call(colors, color)) {
					const colorName = colors[color];
					const renderedAnnotations = await getAnnotationsByColor(
						attachment,
						color
					);
					if (renderedAnnotations) {
						res += `<h3>
	<p style="background-color:${color};">${colorName}</p>
</h3>${renderedAnnotations}`;
					}
				}
			}
			const renderedAnnotations = await getAnnotationsByColor(
				attachment,
				(_annot) => !colors.includes(_annot.annotationColor)
			);
			if (renderedAnnotations) {
				res += `<h2>
	<p>Other Annotations</p>
</h2>${renderedAnnotations}`;
			}
		}
		return res;
	}
} $
```

<hr>
I'd greatly appreciate it if anyone knows how to improve the codes...

In the old script there's one line:

```
await editor.importImages(annotations);
```

**Comment by HPDell** [comment]
The following version works in the latest Zotero. If you encounter problems (e.g., annotations won't show) when using the above version, please try this one.

```js
${await new Promise (async (r) => {
  async function getAnnotation(item) {
    try {
      if (!item || !item.isAnnotation()) {
        return null;
      }
      let json = await Zotero.Annotations.toJSON(item);
      json.id = item.key;
      delete json.key;
      for (let key in json) {
        json[key] = json[key] || "";
      }
      json.tags = json.tags || [];
      return json;
    } catch (e) {
      Zotero.logError(e);
      return null;
    }
  }

  function getEditorInstance(noteId) {
    return Zotero.Notes._editorInstances.find(
      (e) =>
        e._item.id === noteId && !Components.utils.isDeadWrapper(e._iframeWindow)
    );
  }

  async function getAnnotationsByColor(_item, color) {
    const annots = _item.getAnnotations().filter(item => item.annotationColor === color);
    if (annots.length === 0) {
      return {
        html: "",
      };
    }
    let annotations = [];
    for (let annot of annots) {
      const annotJson = await getAnnotation(annot);
      annotJson.attachmentItemID = _item.id;
      annotations.push(annotJson);
    }

    let editor = getEditorInstance(targetNoteItem.id)
    if (!editor) {
      alert("No active note editor detected. Please open workspace.");
      return r("");
    }
    await editor.importImages(annotations);
    return Zotero.EditorInstanceUtilities.serializeAnnotations(annotations);
  }

  const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
    i.isPDFAttachment()
  );

  let res = `<h1>${topItem.getField("title")}</h1>`;

  const colors = {
    "#aaaaaa": "💡Purpose",           // Grey
    "#ffd400": "🖌️General & Images",  // Yellow
    "#ff6666": "❗Attention",         // Red
    "#5fb236": "❓Hypotheses",        // Green
    "#2ea8e5": "📚Theory",            // Blue
    "#a28ae5": "⛏️Method",            // Purple
    "#e56eee": "📝Results",           // Magenta
    "#f19837": "👍Contribution"       // Orange
  }
  for (const attachment of attachments) {
    res += `<h2>📄For Document:${attachment.attachmentFilename}</h2>`;
    for (const color in colors) {
      if (Object.hasOwnProperty.call(colors, color)) {
        const colorName = colors[color];
        const renderedAnnotations = (await getAnnotationsByColor(attachment, color)).html;
        if (renderedAnnotations) {
          res += `<h3><p style="background-color:${color};">${colorName}</p></h3>\n${renderedAnnotations}`;
        }
      }
    }
    const renderedAnnotations = (
      await getAnnotationsByColor(
        attachment,
        (_annot) => !colors.includes(_annot.annotationColor)
      )
    ).html;
    if (renderedAnnotations) {
      res += `<h2><p>Other Annotations</p></h2>\n${renderedAnnotations}`;
    }
  }
  r(res);
})}
```

The problem is caused by variable `editor` which is (possibly) undefined. As the add-on pass the `targetNoteItem` to the template, the real value of `editor` can be extracted from Zotero API through this function

```js
function getEditorInstance(noteId) {
  return Zotero.Notes._editorInstances.find(
    (e) =>
      e._item.id === noteId && !Components.utils.isDeadWrapper(e._iframeWindow)
  );
}
```

Then just call it to get the `editor` before using it

```diff
+   let editor = getEditorInstance(targetNoteItem.id)
    if (!editor) {
      alert("No active note editor detected. Please open workspace.");
      return r("");
    }
    await editor.importImages(annotations);
```

**Reply by butterflyli** [reply]
Thank you very much, this code could run correctly.Also, I wonder what the ItemNotes used in the previous version was replaced with? It makes my other template not working properly

**Reply by HPDell** [reply]
@butterflyli I suppose the `editor` used to be passed from the plugin in previous versions. But now only ``targetNoteItem` is passed. This change caused problems.

**Reply by HPDell** [reply]
I'm afraid the template is not working. Maybe something changed again.

**Reply by windingwind** [reply]
@HPDell Yes, you'll need to update the template. see https://github.com/windingwind/zotero-better-notes/blob/master/docs/about-note-template.md#write-note-template

**Comment by butterflyli** [comment]
Yes,my template is still work well. My zotero’s vesion is 6.0.26.

And my Zotero Better Notes‘ version is 1.0.4.
But I encountered a few times that an empty note was generated.
It might be related to the steps you take to create notes.
After you make annotations on the paper, 
you should first create a new Item note from Zotero Better Notes.

Then, you will get an empty note, please don’t give up.

You should click the gear icon, then choose the ‘Insert template at cursor’

Finally, you will get the right note with  annotations grouped by color


I hope my experience can help you.
 
从 Windows 版邮件发送

发件人: Gino
发送时间: 2023年6月13日 1:40
收件人: windingwind/zotero-better-notes
抄送: butterflyli; Comment
主题: Re: [windingwind/zotero-better-notes] [Item] A slightly modifiedversion of annotation collection by color (Discussion #406)

Your template is still work? Mine is broken in the latest version zotero. There is only a empty note be generated.

**Comment by windingwind** [comment]
`editor` is no longer available in the latest version. There is no promise that an editor of target note exists, so it is removed.

**Reply by npj06** [reply]
how to import this template correctly when I countered a notice “the copied template is invalid”，thanks for your help

**Reply by windingwind** [reply]
https://github.com/windingwind/zotero-better-notes/discussions/406#discussioncomment-6221278

**Comment by windingwind** [comment]
Use this, based on @HPDell : 

One-click to import.

1. Copy the template below.
<details style="text-indent: 2em">
<summary>Click to show the template</summary>

```yaml
name: "[Item] A slightly modified version of annotation collection by color"
content: |-
  // @author LynnXie00, HPDell, and windingwind 
  // @link https://github.com/windingwind/zotero-better-notes/discussions/406
  ${{
  async function getAnnotationsByColor(_item, color) {
    const annots = _item
      .getAnnotations()
      .filter((item) => item.annotationColor === color);
    if (annots.length === 0) {
      return "";
    }
    return Zotero.BetterNotes.api.convert.annotations2html(annots, {
      noteItem: targetNoteItem,
    });
  }

  const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
    i.isPDFAttachment()
  );

  let res = `<h1>${topItem.getField("title")}</h1>`;

  const colors = {
    "#aaaaaa": "💡Purpose", // Grey
    "#ffd400": "🖌️General & Images", // Yellow
    "#ff6666": "❗Attention", // Red
    "#5fb236": "❓Hypotheses", // Green
    "#2ea8e5": "📚Theory", // Blue
    "#a28ae5": "⛏️Method", // Purple
    "#e56eee": "📝Results", // Magenta
    "#f19837": "👍Contribution", // Orange
  };
  for (const attachment of attachments) {
    res += `<h2>📄For Document:${attachment.attachmentFilename}</h2>`;
    for (const color in colors) {
      if (Object.hasOwnProperty.call(colors, color)) {
        const colorName = colors[color];
        const renderedAnnotations = await getAnnotationsByColor(
          attachment,
          color
        );
        if (renderedAnnotations) {
          res += `<h3><p style="background-color:${color};">${colorName}</p></h3>\n${renderedAnnotations}`;
        }
      }
    }
    const renderedAnnotations = await getAnnotationsByColor(
      attachment,
      (_annot) => !colors.includes(_annot.annotationColor)
    );
    if (renderedAnnotations) {
      res += `<h2><p>Other Annotations</p></h2>\n${renderedAnnotations}`;
    }
  }
  return res;
  }}$
```

</details>

2. Goto Zotero menubar, click `Edit`->`New Template from Clipboard`.
3. Click OK.

**Reply by Rubylyp** [reply]
Based on your instructions, I retested, but I still encountered the following error when importing the notes: Template [Item] Color Notes Error: SyntaxError: missing : after property id. Currently, my Zotero version is 6.0.26, and BetterNotes version is 6.7.79.

**Reply by windingwind** [reply]
Only for Better Notes 1.1.3+, Zotero 7, the latest. We have no version 6.7.79.

**Reply by Rubylyp** [reply]
I'm sorry, I checked and confirmed that my "Better Note" version is also 1.0.4. If it's Zotero 7, then it seems like I'll have to wait for a while.

**Reply by Rubylyp** [reply]
@WuG978 I have confirmed that the version of my Better Note is also 1.0.4, Zotero 6. I attempted to pass the above code through
Copy the template
Goto Zotero menubar, click ->.EditNew Template from Clipboard
Then I received the following error message.
![image](https://github.com/windingwind/zotero-better-notes/assets/72732721/2638ed44-6908-419a-baf7-6b07b38b84c1)
Have you encountered similar issues while importing note templates?
I also tried directly copying the template into the note template editor, but I couldn't import it as usual like before.

**Reply by windingwind** [reply]
@WuG978 @Rubylyp 
Use the template in my comment. Click `Click to show the template` and copy it. I gave very, very detailed instructions above.

**Reply by windingwind** [reply]
If you are on Z6 and BN<1.0.0, paste the original template provided in this thread into the editor,
If you are on Z7 and BN>1.1.0, use the new template in this comment (by importing).
Otherwise it won't work.

**Reply by macrospecter** [reply]
I used the new template by importing rather than copying/pasting, but I get this error: "Template [Item] A slightly modified version of annotation collection by color Error: SyntaxError: missing : after property id."

**Reply by HPDell** [reply]
The original template works on Z6 with BN=1.0.4.

**Reply by macrospecter** [reply]
Thanks. I'm on Z6/BN 1.0.4 but for some reason the only thing that worked for me was to copy/paste the amended template (the one you posted May 25, not the original in the first post). And then it worked!

**Reply by npj06** [reply]
Sorry, after reading all the posts about templates, I still can't find the right code to import.While copying the comment area, all the codes are displayed invalid.So I want to ask the author how to solve this problem.

**Reply by windingwind** [reply]
@npj06 

If you are on Z7 and the latest beta of the plugin, copy the import code by clicking this button and follow the steps described above:

<img width="847" alt="image" src="https://github.com/windingwind/zotero-better-notes/assets/33902321/b7d71da0-97bd-4d15-aaaf-9ebf2f514082">

**Reply by AaronLiu0916** [reply]
> The template I quoted can only be add by the template editor, and not added from clipboard. If you want to add it from clipboard, you should modify the template to a certain format. You can refer to the official document. But I haven’t tested it.

Hello, my Zotero and Better Notes have the exact same version as yours, and I mimicked your step to add the template from the template editor but not from the clipboard. However, after I added it, the note generated by Better Notes (regardless of from template or from annotation) went into error code as follows. I wonder whether you have encountered similar problem like this? Thanks!

<img width="678" alt="image" src="https://github.com/windingwind/zotero-better-notes/assets/174075523/01d4848f-5be9-40d0-a519-2edc9a1c05b0">

**Reply by windingwind** [reply]
@AaronLiu0916 please take a screenshot of your template editor window.

**Reply by AaronLiu0916** [reply]
<img width="1539" alt="image" src="https://github.com/windingwind/zotero-better-notes/assets/174075523/8fb01b79-045a-45cd-97fe-fae3ebe6f7af">
<img width="1642" alt="image" src="https://github.com/windingwind/zotero-better-notes/assets/174075523/a8ef000f-928c-4ba4-903c-72c508c96dc4">

This is the template editor window and the note generated for an article. I modified the code from the reply of HPDell of May, 2023 by simply changing the annotation colors:   

const colors = {
   "#ffdd83": "Introduction", //y      
    "#c0797d": "Highlight", //r
    "#91aa9d": "Method",   //g
    "#ffcf79": "Result",  //o
    "#ada2ff": "Question",  //p
    "#a4c9ed": "Word",  //db
    "#c8edec": "Sentence",  //lb
    "#FFD2D2": "Ref"   //pnk
  }

**Reply by AaronLiu0916** [reply]
@windingwind Sorry I did not reply, the answer was put above

**Reply by windingwind** [reply]
What's your Zotero version & plugin version? Also, we need the full template code. (Reply with the template code in the code block so that it won't be recognized as markdown by GitHub)

**Reply by AaronLiu0916** [reply]
@windingwind My Zotero Version is 6.0.37 with Better Notes 1.0.4. In previous replies, I found @Guan9Wu had successfully set up the template with a lower version than required, so I decided to try out.

The whole code was attached below:

```
${await new Promise (async (r) => {
  async function getAnnotation(item) {
    try {
      if (!item || !item.isAnnotation()) {
        return null;
      }
      let json = await Zotero.Annotations.toJSON(item);
      json.id = item.key;
      delete json.key;
      for (let key in json) {
        json[key] = json[key] || "";
      }
      json.tags = json.tags || [];
      return json;
    } catch (e) {
      Zotero.logError(e);
      return null;
    }
  }

  function getEditorInstance(noteId) {
    return Zotero.Notes._editorInstances.find(
      (e) =>
        e._item.id === noteId && !Components.utils.isDeadWrapper(e._iframeWindow)
    );
  }

  async function getAnnotationsByColor(_item, color) {
    const annots = _item.getAnnotations().filter(item => item.annotationColor === color);
    if (annots.length === 0) {
      return {
        html: "",
      };
    }
    let annotations = [];
    for (let annot of annots) {
      const annotJson = await getAnnotation(annot);
      annotJson.attachmentItemID = _item.id;
      annotations.push(annotJson);
    }

    let editor = getEditorInstance(targetNoteItem.id)
    if (!editor) {
      alert("No active note editor detected. Please open workspace.");
      return r("");
    }
    await editor.importImages(annotations);
    return Zotero.EditorInstanceUtilities.serializeAnnotations(annotations);
  }

  const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
    i.isPDFAttachment()
  );

  let res = `<h1>${topItem.getField("title")}</h1>`;

  const colors = {
   "#ffdd83": "Introduction", //y      
    "#c0797d": "Highlight", //r
    "#91aa9d": "Method",   //g
    "#ffcf79": "Result",  //o
    "#ada2ff": "Question",  //p
    "#a4c9ed": "Word",  //db
    "#c8edec": "Sentence",  //lb
    "#FFD2D2": "Ref"   //pnk
  }
  for (const attachment of attachments) {
    res += `<h2>📄For Document:${attachment.attachmentFilename}</h2>`;
    for (const color in colors) {
      if (Object.hasOwnProperty.call(colors, color)) {
        const colorName = colors[color];
        const renderedAnnotations = (await getAnnotationsByColor(attachment, color)).html;
        if (renderedAnnotations) {
          res += `<h3><p style="background-color:${color};">${colorName}</p></h3>\n${renderedAnnotations}`;
        }
      }
    }
    const renderedAnnotations = (
      await getAnnotationsByColor(
        attachment,
        (_annot) => !colors.includes(_annot.annotationColor)
      )
    ).html;
    if (renderedAnnotations) {
      res += `<h2><p>Other Annotations</p></h2>\n${renderedAnnotations}`;
    }
  }
  r(res);
})}
```

**Comment by yuanyuanfka** [comment]
怎么下线这个插件啊友友们

**Comment by anyu543937321** [comment]
This note template is very useful, and I've modified the colors of the annotations to match my reading preferences. However, I'm wondering if it's possible to also make the text color in the notes consistent with the annotation colors.

// @author LynnXie00, HPDell, and windingwind 
// @link https://github.com/windingwind/zotero-better-notes/discussions/406
${{
async function getAnnotationsByColor(_item, color) {
  const annots = _item
    .getAnnotations()
    .filter((item) => item.annotationColor === color);
  if (annots.length === 0) {
    return "";
  }
  return Zotero.BetterNotes.api.convert.annotations2html(annots, {
    noteItem: targetNoteItem,
  });
}

const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
  i.isPDFAttachment()
);

let res = `<h1>${topItem.getField("title")}</h1>`;

const colors = {
  
   "#ffd400": "🖌️Overview", // Yellow
   "#e56eee": "❓Key Issues", // Magenta
   "#f19837": "🔨Methodology", // Orange
   "#2ea8e5": "🏡Framework", // Blue
   "#ff6666": "🤔Viewpoint", // Red
   "#aaaaaa": "📝Conclusion", // Grey
   "#5fb236": "📚Reference", // Green
   "#a28ae5": "🆕New words", // Purple
};
for (const attachment of attachments) {
  res += `<h2>📄For Document:${attachment.attachmentFilename}</h2>`;
  for (const color in colors) {
    if (Object.hasOwnProperty.call(colors, color)) {
      const colorName = colors[color];
      const renderedAnnotations = await getAnnotationsByColor(
        attachment,
        color
      );
      if (renderedAnnotations) {
        res += `<h3><p style="background-color:${color};">${colorName}</p></h3>\n${renderedAnnotations}`;
      }
    }
  }
  const renderedAnnotations = await getAnnotationsByColor(
    attachment,
    (_annot) => !colors.includes(_annot.annotationColor)
  );
  if (renderedAnnotations) {
    res += `<h2><p>Other Annotations</p></h2>\n${renderedAnnotations}`;
  }
}
return res;
}}$

**Reply by ititiy** [reply]
可以直接右上角设置了现在

**Comment by DocRuzzy** [comment]
Thanks so much for this template!   I modified it for my own color coding (which also includes colors used by the Boox Neoreader)
```
// @author LynnXie00, HPDell, and windingwind 
// @link https://github.com/windingwind/zotero-better-notes/discussions/406
${{
async function getAnnotationsByColor(_item, color) {
  const annots = _item
    .getAnnotations()
    .filter((item) => item.annotationColor === color);
  if (annots.length === 0) {
    return "";
  }
  return Zotero.BetterNotes.api.convert.annotations2html(annots, {
    noteItem: targetNoteItem,
  });
}

const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
  i.isPDFAttachment()
);

let res = `<h1>${topItem.getField("title")}</h1>`;

const colors = {
  "#ffd400": "⭐ Interesting point, Facts, Examples, References", // Yellow
  "#FFF0FF00": "⭐ Interesting point, Facts, Examples, References", // Yellow neoreader
  "#5fb236": "🧪 Main findings, Results, Conclusions", // Green
  "#FF00B036": "🧪 Main findings, Results, Conclusions", // Green neoreader
  "#f19837": "❓Assumptions, Questions, Goals, Problems, challenges", // Orange
  "#FFFFAA00": "❓Assumptions, Questions, Goals, Problems, challenges", // Orange neoreader
  "#2ea8e5": "⚗️Experimental Details or Methods", // Blue
  "#FF00F0FF": "⚗️Experimental Details or Methods", // Blue neoreader
  "#a28ae5": "♐ Supporting Argument or Example", // Purple
  "#FF9338BE": "♐ Supporting Argument or Example", // Purple neoreader
  "#ff6666": "⛔ Contrary/Null Example or Argument", // Red
  "#FFFF6163": "⛔ Contrary/Null Example or Argument", // Red neoreader
  "#aaaaaa": "🗃️ Vocabulary, Names, Dates, Definitions", //gray
  "#FFC0C0C0": "🗃️ Vocabulary, Names, Dates, Definitions", //gray neoreader
  "#e56eee": "🧠 Background information, Prerequisites", // Magenta
  "#FFEE00FF": "🧠 Background information, Prerequisites", // Magenta neoreader
  "#FF00AAFF": "🎇 applications", //sky blue neoreader
  "#FFFF4400": "⚠️ Disagree with author", //orange red neoreader
  "#FF808080": "Notable references", //medium gray neoreader
  "#FFFFFFFF": "🤐 hidden", //white neoreader
};
for (const attachment of attachments) {
  res += `<h2>📄For Document:${attachment.attachmentFilename}</h2>`;
  for (const color in colors) {
    if (Object.hasOwnProperty.call(colors, color)) {
      const colorName = colors[color];
      const renderedAnnotations = await getAnnotationsByColor(
        attachment,
        color
      );
      if (renderedAnnotations) {
        res += `<h3><p style="background-color:${color};">${colorName}</p></h3>\n${renderedAnnotations}`;
      }
    }
  }
  const renderedAnnotations = await getAnnotationsByColor(
    attachment,
    (_annot) => !colors.includes(_annot.annotationColor)
  );
  if (renderedAnnotations) {
    res += `<h2><p>Other Annotations</p></h2>\n${renderedAnnotations}`;
  }
}
return res;
}}$
```

**Reply by Geo123abc** [reply]
thanks, can you tell me how to modify zotero's default highlight color?

**Comment by kepbod** [comment]
每次如果annotation 更新了，note 也更着更新或者有快捷键点一下就更新就好了。

**Reply by kepbod** [reply]
Or we can add an action to update the note (maybe delete the old one and add a new one)?

**Reply by danieltregeagle** [reply]
> Hi, great template! But when adding an annotation to notes, is it possible to automatically quote the highlighted text and place the note of the highlighted text in a new paragraph?
> 
> <img alt="problem" width="1025" src="https://private-user-images.githubusercontent.com/168875915/327900674-4789df9d-cf29-45de-85d8-3452221a66e5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5MDgwMjYsIm5iZiI6MTcxODkwNzcyNiwicGF0aCI6Ii8xNjg4NzU5MTUvMzI3OTAwNjc0LTQ3ODlkZjlkLWNmMjktNDVkZS04NWQ4LTM0NTIyMjFhNjZlNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyMFQxODIyMDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YjIzOGZmN2M5Yjg0ZDg1ZjcwZThmMjk5MGM4ZWRiZDlmN2NmMDQyYWE3MDE4NTVkMmRhYjMzNmFhMDI1Yzk3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.GQepQuigZj9JQkUh-t48cQEKKVupP43-fPAcaNlIzt4">

You can modify the appearance of pdf highlights and notes using [note templates](https://www.zotero.org/support/note_templates). [See this thread](https://forums.zotero.org/discussion/95600/adding-annotation-tags-to-notes) for an example.

**Reply by huberdf** [reply]
great idea，how to achieve it?

**Comment by oliverkuehle** [comment]
Seems not to be working with the recent zotero versions.
I would greatly appreciate if somebody could share a working fix for the zotero 7 beta 🙏

**Reply by fredericky123** [reply]
蹲个适配zotero7的版本

**Comment by fredericky123** [comment]
文末自动生成的这个横线的颜色应该怎么修改呢
<img width="349" alt="image" src="https://github.com/user-attachments/assets/e79724d0-2aee-4c4d-9924-b6a3c37acadc">

**Comment by VincentCusson** [comment]
> **2024-10-10
> 
> Guys, I updated the scripts. With the Zotero 7 and Better Notes, for some reason, the script can't import annotations that contain images anymore. (I'm not a programmer and that's all I can do lol.... So you probably need to import images into Zotero notes manually. But the other part works fine for me. **
> 
> This code is modified based on later versions in this thread [#406 (comment)](https://github.com/windingwind/zotero-better-notes/discussions/406#discussioncomment-6887168)

I get the error `SyntaxError: unexpected token: identifier`  when saving and using the template.

Zotero 7.0.7 and Better Notes 2.0.17

Any idea what could cause that ?

**Reply by LynnXie00** [reply]
That indicates some grammar errors... you might have deleted something when editing the script...  I am using this exact one on Zotero 7 and the latest BN, it works perfectly (just no image import anymore)


```
${{
async function getAnnotationsByColor(_item, color) {
const annots = _item
.getAnnotations()
.filter((item) => item.annotationColor === color);
if (annots.length === 0) {
return "";
}
return Zotero.BetterNotes.api.convert.annotations2html(annots, {
noteItem: targetNoteItem,
});
}

const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
i.isPDFAttachment()
);

const creators = topItem.getCreators();
let content = "";
const year = topItem.getField("year");
if (creators.length === 0) {
  content = "no author";
} else if (creators.length === 1) {
  content = `${creators[0].lastName}, ${year}`;
} else {
  content = `${creators[0].lastName} etal., ${year}`;
}

let res = `<h1>${content}</h1> <h2>${topItem.getField("title")}</h2>`;

const colors = {
"#aaaaaa": "💡 Purpose", // Grey
"#fd7b41": "🔍 Research Gap", // Redish Orange (customized)
"#2ea8e5": "📚 Theory", // Blue
"#5fb236": "❗Hypotheses", // Green
"#a28ae5": "⛏️ Method", // Purple
"#e56eee": "📝 Results", // Magenta
"#f19837": "👍 Contribution", // Orange
"#ff6666": "⁉️ Attention & Limitation", // Red
"#34eb86": "🕞 Future Research", // Lighter Green(customized)

};
for (const attachment of attachments) {
res += `<h2>📄For Document:${attachment.attachmentFilename}</h2>`;
for (const color in colors) {
if (Object.hasOwnProperty.call(colors, color)) {
const colorName = colors[color];
const renderedAnnotations = await getAnnotationsByColor(
attachment,
color
);
if (renderedAnnotations) {
res += `<h3>
	<p style="background-color:${color};">${colorName}</p>
</h3>${renderedAnnotations}`;
}
}
}
const renderedAnnotations = await getAnnotationsByColor(
attachment,
(_annot) => !colors.includes(_annot.annotationColor)
);
if (renderedAnnotations) {
res += `<h2>
	<p>Other Annotations</p>
</h2>${renderedAnnotations}`;
}
}
return res;
}}$

```

**Reply by windingwind** [reply]
> just no image import anymore

Do you mean the template won't embed image annotations' image into the note? If so, it would be best to report this as a bug in a new issue.

**Comment by macrospecter** [comment]
Still loving this template. I'm wondering if there might be a way to edit it, though, so that footnote numbers do not show in the quoted annotation text? Or at least to delete any numbers that immediately follow text/punctuation without a space? @windingwind 

Unfortunately this is a bit above my technical abilities.

**Reply by windingwind** [reply]
Please do not at me - contact the OP who provided the template

**Reply by macrospecter** [reply]
Apologies!

**Comment by mmkeefe** [comment]
I've just tried to import this template into zotero and get a message that the copied template is invalid. I'm using Zotero 7.0.11 and BN 2.2.5. Does anyone else have it working in these versions? Thank you!

**Comment by re-re-rea** [comment]
Hello! 
I edited the script based on my color codes and tried to use AI since I have zero coding knowledge for making the quotes bold while keeping the rest normal. It works perfectly on my Zotero. However when a friend of mine tried it on his Zotero, it didnt work. Can anyone let me know what's the issue with the code and how can I make it work in other Zoteros? I'm planning to use it with my students. Any help would be much appreciated!

<img width="820" height="601" alt="image" src="https://github.com/user-attachments/assets/ca26d6e9-f44a-4b25-9e85-830db34a4fe1" />

```// @author LynnXie00, HPDell, and windingwind 
// @link https://github.com/windingwind/zotero-better-notes/discussions/406
${{
async function getAnnotationsByColor(_item, color) {
  const annots = _item
    .getAnnotations()
    .filter((item) => item.annotationColor === color);
  if (annots.length === 0) {
    return "";
  }
  const html = await Zotero.BetterNotes.api.convert.annotations2html(annots, {
    noteItem: targetNoteItem,
  });
  
  // Make the highlight (annotation text) and citation bold
  // Target the <span class="highlight"> and <span class="citation"> tags
  let modifiedHtml = html.replace(
    /(<span class="highlight"[^>]*>.*?<\/span>\s*<span class="citation"[^>]*>.*?<\/span>)/gs,
    '<strong>$1</strong>'
  );
  
  return modifiedHtml;
}

const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
  i.isPDFAttachment()
);
let res = `<h1>${topItem.getField("title")}</h1>`;
const colors = {
  "#ff6666": "🚨Primary", // Red
  "#f19837": "🛎️Secondary", // Orange
  "#2ea8e5": "📚Theory", // Blue
  "#e56eee": "📝Key Terms", // Magenta
  "#a28ae5": "⛏️Method", // Purple
  "#5fb236": "🔎Results", // Green
  "#ffd400": "📜Resources", // Yellow
  "#aaaaaa": "📈Images & Charts & Shapes & Data", // Grey
};
for (const attachment of attachments) {
  res += `<h2>📄For Document:${attachment.attachmentFilename}</h2>`;
  for (const color in colors) {
    if (Object.hasOwnProperty.call(colors, color)) {
      const colorName = colors[color];
      const renderedAnnotations = await getAnnotationsByColor(
        attachment,
        color
      );
      if (renderedAnnotations) {
        res += `<h3><p style="background-color:${color};">${colorName}</p></h3>\n${renderedAnnotations}`;
      }
    }
  }
  const renderedAnnotations = await getAnnotationsByColor(
    attachment,
    (_annot) => !colors.includes(_annot.annotationColor)
  );
  if (renderedAnnotations) {
    res += `<h2><p>Other Annotations</p></h2>\n${renderedAnnotations}`;
  }
}
return res;
}}$```
